### PR TITLE
Add opt-in payload integrity protection to cache backends

### PR DESCRIPTION
**Pull request by iliaal**
### Motivation

`Cache::get()` unserializes whatever bytes the cache storage returns. Any principal that can write cache entries — a shared memcached/redis server without authentication, a neighboring application on a shared host, or cache poisoning through user-keyed entries — can instantiate arbitrary framework objects on the next read of that entry.

Because `Dependency` objects ride inside the same serialized payload as the value, this is worse than generic POP-gadget exposure: a poisoned entry containing an `ExpressionDependency` evaluates its expression via `eval()` during the dependency check, and a `CallbackDependency` invokes its restored callback. Cache write access therefore becomes deterministic remote code execution using only first-party, always-loaded classes — no third-party gadget chain required. (This is the same bug class as CVE-2020-15148, but reached through the framework's own default read path rather than application-level `unserialize()`.)

Demonstrated end-to-end on the current master with a payload of the shape

```
serialize(['pwned', $dep]) // where $dep is ExpressionDependency{expression: "system('...')"}
```

written directly to the backend storage: the next `$cache->get()` executes the expression.

### Changes

Introduce an opt-in `Cache::$integrityKey`. When set:

* payloads written through `set()`, `add()`, `multiSet()` and `multiAdd()` are prefixed with an HMAC-SHA256 signature;
* `get()` and `multiGet()` treat entries whose signature is missing or invalid like missing entries instead of unserializing them, which also covers tampered bytes and foreign unsigned payloads.

The option is off by default, so existing behavior (including `serializer => false` backends such as `ArrayCache`) is unchanged; enabling it adds one HMAC computation per access and invalidates entries written beforehand. The guide already documents that cache backends must stay write-protected; this option provides integrity enforcement for deployments where that cannot be guaranteed.

### Tests

New `tests/framework/caching/CacheIntegrityTest.php` covers: round-trip reads/writes with a key enabled, dependency evaluation and invalidation over signed payloads, byte-tampered payloads treated as misses, unsigned payloads rejected, a poisoned `ExpressionDependency` payload no longer evaluating, custom serializers combined with the signature, and unchanged default behavior when the key is not set.

### CODE REVIEW

File: framework/caching/Cache.php [hunk 1]
```diff
@@ -135,6 +148,11 @@ public function get($key)
         $value = $this->getValue($key);
         if ($value === false || $this->serializer === false) {
             return $value;
+        }
+
+        $value = $this->verifyPayload($value, $key);
```
Comment (terabytesoftw) [inline framework/caching/Cache.php:153]: [P1] Verify before backend decoding

`getValue()` has already invoked backend-specific decoding when verification occurs here. `ApcCache` delegates to `apcu_fetch()`, which can reconstruct objects and invoke `__wakeup()` before returning. A local proof printed `WAKEUP_BEFORE_HMAC` before `get()` returned `false`. Verify raw bytes before backend deserialization or explicitly reject unsupported native typed backends.

File: framework/caching/Cache.php [hunk 2]
```diff
@@ -365,6 +388,67 @@ private function prepareCacheData($items, $dependency)
...
+     * the value untouched in that case.
+     *
+     * @param mixed $value the serialized payload.
+     * @param string $key the normalized cache key the payload is stored under.
+     * @return mixed the signed payload.
+     * @since 2.0.56
+     */
+    protected function signPayload($value, $key)
+    {
+        if ($this->integrityKey === null || $this->serializer === false || !is_string($value)) {
```
Comment (terabytesoftw) [inline framework/caching/Cache.php:405]: [P1] Validate integrity keys

Only `null` disables protection, so `''` enables HMAC with a publicly known empty key. A locally forged payload signed with that key executed an injected `ExpressionDependency`. Reject non-string, empty, and insufficiently strong keys with `InvalidConfigException`.

File: framework/caching/Cache.php [hunk 3]
```diff
@@ -365,6 +388,67 @@ private function prepareCacheData($items, $dependency)
...
+     * cannot be copied to a different key.
+     * The option has no effect when [[serializer]] is `false`, so this method leaves
+     * the value untouched in that case.
+     *
+     * @param mixed $value the serialized payload.
+     * @param string $key the normalized cache key the payload is stored under.
+     * @return mixed the signed payload.
+     * @since 2.0.56
+     */
+    protected function signPayload($value, $key)
```
Comment (terabytesoftw) [inline framework/caching/Cache.php:403]: [P1] Avoid protected-name BC break

Adding these protected helpers to an extensible base class can make existing cache subclasses fatal when they already define a homonymous method with another signature or visibility. These are internal helpers rather than extension points; make both helpers private and preferably use uniquely internal names.

File: framework/caching/Cache.php [hunk 4]
```diff
@@ -252,6 +273,7 @@ public function set($key, $value, $duration = null, $dependency = null)
             $value = call_user_func($this->serializer[0], [$value, $dependency]);
         }
         $key = $this->buildKey($key);
+        $value = $this->signPayload($value, $key);
```
Comment (terabytesoftw) [inline framework/caching/Cache.php:276]: [P2] Reject unsupported serializer output

Non-string custom serializer results are stored unsigned here but are rejected by `verifyPayload()`. With `ArrayCache` and an identity serializer, `set()` returns `true` while `get()` returns `false`. Require a string result and fail before writing, or provide an authenticated representation for mixed values.

File: framework/caching/Cache.php [hunk 5]
```diff
@@ -365,6 +388,67 @@ private function prepareCacheData($items, $dependency)
...
+     * @param string $key the normalized cache key the payload is expected under.
+     * @return mixed the unsigned payload or `false` on verification failure.
+     * @since 2.0.56
+     */
+    protected function verifyPayload($value, $key)
+    {
+        if ($this->integrityKey === null) {
+            return $value;
+        }
+        if (!is_string($value) || StringHelper::byteLength($value) <= 64) {
```
Comment (terabytesoftw) [inline framework/caching/Cache.php:426]: [P2] Accept an empty signed payload

A valid signature for an empty serialized payload is exactly 64 bytes, so this condition rejects data written successfully by `signPayload()`. Use `< 64` and add a custom-serializer regression test for an empty payload.

File: tests/framework/caching/CacheIntegrityTest.php [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+        $cache = new IntegrityTestCache();
+        $cache->integrityKey = 'unit-test-integrity-key-0123456789abcdef';
+
+        return $cache;
+    }
+
+    public function testGetSetRoundTrip()
+    {
+        $cache = $this->createCache();
+        $cache->set('k', ['hello' => 'world']);
```
Comment (terabytesoftw) [inline tests/framework/caching/CacheIntegrityTest.php:79]: [P2] Cover every write path

The integrity-enabled round trip only exercises `set()`. Add coverage for `add()`, `multiSet()`, and `multiAdd()`, since those modified paths are part of the security guarantee.

File: tests/framework/caching/CacheIntegrityTest.php [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+
+    public function testDependencyRoundTrip()
+    {
+        $cache = $this->createCache();
+        $dependency = new MutableDependency();
+        $cache->set('k', 'value', 0, $dependency);
+        $this->assertSame('value', $cache->get('k'));
+
+        $dependency->evaluateDependency($cache);
+        MutableDependency::$sourceValue = 2;
```
Comment (terabytesoftw) [inline tests/framework/caching/CacheIntegrityTest.php:93]: [P3] Reset shared test state

This static value is never restored. Running this dependency scenario twice in the same process leaves the value at `2`, so the second execution no longer detects a change. Reset it during setup or teardown.

**Review by terabytesoftw — CHANGES_REQUESTED** [review]
Add an entry to `framework/CHANGELOG.md` for the new public `Cache::$integrityKey` feature.

The inline comments below include blocking security and backward-compatibility issues that should be resolved before approval.
